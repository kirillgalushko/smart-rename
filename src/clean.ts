export interface CleanOptions {
  blacklist: (string | RegExp)[];
  removeSpaces?: boolean;
}

export const clean = (input: string, options: CleanOptions) => {
  const { blacklist, removeSpaces } = options;

  let result = input;

  blacklist.forEach((pattern) => {
    result = result.replace(new RegExp(pattern, 'g'), '');
  });

  if (removeSpaces) {
    result = result.replace(/\s+/g, '').trim();
  }

  return result;
};

export default clean;
