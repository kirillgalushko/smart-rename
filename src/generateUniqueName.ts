import log from 'log';

export type FormatCounterFn = (name: string, counter: number) => string;

export const defaultFormatCounter: FormatCounterFn = (name, counter) =>
  `${name} (${counter})`;

export const generateUniqueName = (
  name: string,
  uniqNamesSet: Set<string>,
  formatCounter: FormatCounterFn = defaultFormatCounter
) => {
  log.info(`Attempting to generate uniq name for: ${name}`);
  let uniqueName = name;
  let counter = 2;

  while (uniqNamesSet.has(uniqueName)) {
    log.info(`${uniqueName} is already taken, increase counter`);
    uniqueName = formatCounter(name, counter);
    counter++;
  }

  log.info(`Uniq name for ${name} is ${uniqueName}`);
  return uniqueName;
};
