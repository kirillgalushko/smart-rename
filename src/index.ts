import { ensureDir, readdir } from 'fs-extra';
import cleanFilename, { type CleanOptions } from './clean';
import path from 'path';
import log from 'log';

export const RESULT_DIR = 'rename_result';

export const createResultDirectory = async (inputPath: string) => {
  const resultDirPath = path.join(inputPath, RESULT_DIR);
  log.info(`Attempting to create result directory: ${resultDirPath}`);

  try {
    await ensureDir(resultDirPath);
    log.info(`Result directory created successfully: ${resultDirPath}`);
  } catch (error) {
    log.error(`Error creating result directory: ${resultDirPath}`, error);
    throw error;
  }
};

export const getFilesInDirectory = async (
  dirPath: string
): Promise<string[]> => {
  log.info(`Attempting to get files in directory: ${dirPath}`);
  try {
    const files = await readdir(dirPath, { withFileTypes: true });
    const filePaths = files
      .filter((file) => file.isFile())
      .map((file) => path.join(dirPath, file.name));

    log.info(`Files found in directory: ${filePaths.length}`);
    return filePaths;
  } catch (error) {
    log.error(`Error reading directory: ${dirPath}`, error);
    throw error;
  }
};

export const getCleanedFilesMap = (
  filePaths: string[],
  clean: (input: string) => string
): Map<string, string> => {
  log.info(`Attempting to create a map of cleaned filenames`);
  try {
    const cleanedPaths = new Map<string, string>();
    for (const oldPath of filePaths) {
      const extname = path.extname(oldPath);
      const baseName = path.basename(oldPath, extname);
      const newBaseName = clean(baseName);
      const newPath = path.join(path.dirname(oldPath), newBaseName + extname);
      cleanedPaths.set(oldPath, newPath);
    }
    log.info(`Cleaned filenames map size: ${cleanedPaths.size}`);
    return cleanedPaths;
  } catch (error) {
    log.error(`Error while creating a map of cleaned filenames`, error);
    throw error;
  }
};

export const rename = async (inputPath: string) => {
  log.info(`Starting the renaming process for inputPath: ${inputPath}`);

  try {
    await createResultDirectory(inputPath);
    log.info('Renaming process completed successfully');
  } catch (error) {
    log.error('Renaming process failed', error);
  }
};
