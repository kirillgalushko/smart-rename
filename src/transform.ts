export interface TransformOptions {
  patternsToRemove: (string | RegExp)[];
  removeSpaces?: boolean;
}

export const transform = (input: string, options: TransformOptions) => {
  const { patternsToRemove, removeSpaces } = options;

  let result = input;

  patternsToRemove.forEach((pattern) => {
    result = result.replace(new RegExp(pattern, 'g'), '');
  });

  if (removeSpaces) {
    result = result.replace(/\s+/g, '').trim();
  }

  return result;
};

export default transform;
