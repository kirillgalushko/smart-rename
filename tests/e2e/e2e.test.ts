import { describe, it, expect, beforeEach, afterAll } from 'vitest';
import fs from 'fs-extra';
import path, { basename } from 'path';
import { smartRename } from '../../src/index';

const TEST_RESULT_DIR_NAME = 'test_result';

const patternsToRemove = [
  /-svgrepo-com/g,
  /-logo/g,
  /-color/g,
  /-icon/g,
  /_\- logo/g,
  /\d{1,}x\d{1,}/g,
  /\(.*?\)/g,
  /_.*/g,
  /id[^\s]*/g,
  /^-|-$/g,
  /audioblocks-/g,
];

describe('e2e', () => {
  const mocksDir = path.resolve(__dirname, 'mocks');
  const testResultDir = path.resolve(__dirname, TEST_RESULT_DIR_NAME);

  beforeEach(async () => {
    await new Promise((resolve) => setTimeout(resolve, 100));
    await fs.remove(testResultDir);
    await fs.copy(mocksDir, testResultDir);
  });

  afterAll(async () => {
    await new Promise((resolve) => setTimeout(resolve, 100));
    await fs.remove(testResultDir);
  });

  it('should execute basic functionality', async () => {
    await smartRename(testResultDir, {
      patternsToRemove,
      removeSpaces: true,
    });
    const resultFiles = await fs.readdir(testResultDir);
    expect(resultFiles).toMatchSnapshot();
  });

  it('should correctly transform filenames', async () => {
    await smartRename(testResultDir, {
      transform: (filename) => {
        const prefix = 'prefix-';
        const postfix = '-postfix';
        return prefix + filename + postfix;
      },
    });
    const resultFiles = await fs.readdir(testResultDir);
    expect(resultFiles).toMatchSnapshot();
  });

  it('should correctly transform filenames after removing patterns', async () => {
    await smartRename(testResultDir, {
      patternsToRemove,
      removeSpaces: true,
      transform: (filename) => {
        const prefix = 'prefix-';
        const postfix = '-postfix';
        return prefix + filename + postfix;
      },
    });
    const resultFiles = await fs.readdir(testResultDir);
    expect(resultFiles).toMatchSnapshot();
  });

  it('should correctly add custom counter', async () => {
    await smartRename(testResultDir, {
      patternsToRemove,
      removeSpaces: true,
      formatCounter: (name, counter) => {
        return `${name}__${counter}`;
      },
    });
    const resultFiles = await fs.readdir(testResultDir);
    expect(resultFiles).toMatchSnapshot();
  });

  it('should correctly handle multiple run', async () => {
    const run = async () =>
      smartRename(testResultDir, {
        patternsToRemove,
        removeSpaces: true,
      });
    await run();
    await run();
    await run();
    const resultFiles = await fs.readdir(testResultDir);
    expect(resultFiles).toMatchSnapshot();
  });
});
