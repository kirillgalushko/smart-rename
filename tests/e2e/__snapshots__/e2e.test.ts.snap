// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`e2e > should correctly add custom counter 1`] = `
[
  "digital-ocean.svg",
  "download",
  "fatal-error-2.mp3",
  "gameloft.svg",
  "intel.svg",
  "intel.txt",
  "light-switch.mp3",
  "login.txt",
  "mattermost.svg",
  "mattermost__2.svg",
  "mouse-click-computer.txt",
  "notification-message-light-interface-sound-2.mp3",
  "pleasant-interface-alert-2-message-hint-ding-message-hint-ding.mp3",
]
`;

exports[`e2e > should correctly handle multiple run 1`] = `
[
  "digital-ocean.svg",
  "download",
  "fatal-error-2.mp3",
  "gameloft.svg",
  "intel.svg",
  "intel.txt",
  "light-switch.mp3",
  "login.txt",
  "mattermost (2).svg",
  "mattermost.svg",
  "mouse-click-computer.txt",
  "notification-message-light-interface-sound-2.mp3",
  "pleasant-interface-alert-2-message-hint-ding-message-hint-ding.mp3",
]
`;

exports[`e2e > should correctly transform filenames 1`] = `
[
  "prefix-audioblocks-fatal-error-2_SY8xTgf8ADU_NWM-postfix.mp3",
  "prefix-audioblocks-notification-message-light-interface-sound-2_SFIlEpWL0vU_NWM-postfix.mp3",
  "prefix-audioblocks-pleasant-interface-alert-2-message-hint-ding-message-hint-ding_Ht4sRWUAvU_NWM-postfix.mp3",
  "prefix-digital-ocean-svgrepo-com (1)-postfix.svg",
  "prefix-download-id4knLKYsV-1732567273343-postfix",
  "prefix-gameloft-postfix.svg",
  "prefix-intel (1)-postfix.txt",
  "prefix-intel_- logo-1000x660-postfix.svg",
  "prefix-light-switch_NWM-postfix.mp3",
  "prefix-login_Msd12dsRG3-postfix.txt",
  "prefix-mattermost-icon-svgrepo-com-postfix.svg",
  "prefix-mattermost-svgrepo-com-postfix.svg",
  "prefix-mouse-click-computer_Gkjr_nNu_NWM-postfix.txt",
]
`;

exports[`e2e > should correctly transform filenames after removing patterns 1`] = `
[
  "prefix-digital-ocean-postfix.svg",
  "prefix-download-postfix",
  "prefix-fatal-error-2-postfix.mp3",
  "prefix-gameloft-postfix.svg",
  "prefix-intel-postfix.svg",
  "prefix-intel-postfix.txt",
  "prefix-light-switch-postfix.mp3",
  "prefix-login-postfix.txt",
  "prefix-mattermost-postfix (2).svg",
  "prefix-mattermost-postfix.svg",
  "prefix-mouse-click-computer-postfix.txt",
  "prefix-notification-message-light-interface-sound-2-postfix.mp3",
  "prefix-pleasant-interface-alert-2-message-hint-ding-message-hint-ding-postfix.mp3",
]
`;

exports[`e2e > should execute basic functionality 1`] = `
[
  "digital-ocean.svg",
  "download",
  "fatal-error-2.mp3",
  "gameloft.svg",
  "intel.svg",
  "intel.txt",
  "light-switch.mp3",
  "login.txt",
  "mattermost (2).svg",
  "mattermost.svg",
  "mouse-click-computer.txt",
  "notification-message-light-interface-sound-2.mp3",
  "pleasant-interface-alert-2-message-hint-ding-message-hint-ding.mp3",
]
`;
